### ISSUE TITLE
Enhancement: Remove getSourceFileOfNode from type-utils

### ISSUE THREAD

**Issue by y-hsgw** [typescript-eslint/typescript-eslint#9797]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

type-utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

It appears that the getSourceFileOfNode function is not being used. I think it's safe to delete.
https://github.com/typescript-eslint/typescript-eslint/blob/a8f91e3b84e0afcb4c08b23fde2596517eb9810e/packages/type-utils/src/getSourceFileOfNode.ts#L6

**Comment by JoshuaKGoldberg** [comment]
Indeed, good catch. The only problem is, this would be a breaking change as it's publicly exported now... and we're a long time away from a v9/v10. Marking as blocked for now.

**Comment by JoshuaKGoldberg** [comment]
Oh actually, in the meantime we can just mark it as `@deprecated`. That's not a breaking change. And we generally do a full sweep of deprecated code in majors.
